src/functions.php

```diff
@@ -558,33 +558,42 @@ function extract_session_from_options(array $options): ?Session
  */
 function extract_read_preference_from_options(array $options): ?ReadPreference
 {
-    if (! isset($options['readPreference']) || ! $options['readPreference'] instanceof ReadPreference) {
-        return null;
+    if (isset($options['readPreference']) && $options['readPreference'] instanceof ReadPreference) {
+        return $options['readPreference'];
     }
 
-    return $options['readPreference'];
+    return null;
 }
 
 /**
- * Performs server selection, respecting the readPreference and session options
- * (if given)
+ * Performs server selection, respecting the readPreference and session options.
+ *
+ * The pinned server for an active transaction takes priority, followed by an
+ * operation-level read preference, followed by an active transaction's read
+ * preference, followed by a primary read preference.
  *
  * @internal
  */
 function select_server(Manager $manager, array $options): Server
 {
     $session = extract_session_from_options($options);
-    $server = $session instanceof Session ? $session->getServer() : null;
-    if ($server !== null) {
-        return $server;
+
+    // Pinned server for an active transaction takes priority
+    if ($session instanceof Session && $session->getServer() instanceof Server) {
+        return $session->getServer();
```

Comment:
Note that the previous logic of storing the result of `$session->getServer()` in its own variable was to work around the Psalm error. `$session->getServer()` may return a different result on subsequent invocations (as unlikely it is in a single-threaded context). Storing the result in a variable and checking that variable ensures that we're always returning a `Server` instance.

Reply:
Thanks. I obviously forgot to run `psalm` locally before publishing the PR :smile: